Add intermediate_value_type column to represent inf/-inf on RDBStorage.

c-bata opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull requests after they get two or more approvals. -->

## Motivation
<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->
Refs #3206 
Follow-up https://github.com/optuna/optuna/pull/3238 and https://github.com/optuna/optuna/pull/3327

To store `inf` and `-inf` as an intermediate_value, #3228 introduces the logic that clips intermediate_value like `np.clip(intermediate_value, np.finfo(np.float32).min, np.finfo(np.float32).max)`. But the logic was broken at #3327 since we changed the floating point precision to double precision numbers.

## Description of the changes
<!-- Describe the changes in this PR. -->

This PR adds a new enum column that makes intermediate_value able to represent `inf` and `-inf`.

I confirmed the migration script by using himkt 's optuna-2e. The corresponding branch is:
https://github.com/c-bata/optuna-e2e/tree/optuna-3564

c-bata reviewed (commented):
@toshihikoyanase @contramundum53 Could you review this PR?

toshihikoyanase commented on optuna/storages/_rdb/storage.py:
```diff
@@ -814,6 +816,22 @@ def _lift_numerical_limit(value: Optional[float]) -> float:
         return value
 
+    @staticmethod
+    def _convert_intermediate_value_representation(
+        intermediate_value_model: models.TrialIntermediateValueModel,
+    ) -> float:
```
The name of this function might be a bit ambiguous since it does not indicate how this function convert the representation. I don't have any particular ideas, but we may be able to name after keywords like float type, such as `_revert_float_value_of_intermediate_value`. Or, `_revert_intermediate_value` for simplicity.

c-bata replied:
Yes, that makes sense. Especially `_revert_float_value_of_intermediate_value` sounds better. Or we may use:

* `_construct_float_from_intermediate_values_model` 
* `_to_intermediate_value_from_sqlalchemy_model`

c-bata replied:
Yes, that makes sense. Especially `_revert_float_value_of_intermediate_value` sounds better. Or we may use:

* `_construct_float_from_intermediate_values_model` 
* `_to_intermediate_value_from_sqlalchemy_model`

toshihikoyanase replied:
Thank you for suggesting the candidates.

> `_to_intermediate_value_from_sqlalchemy_model`

This one is easier to understand  for me, but all of them are descriptive. I guess this method is private and we can rename it if necessary, so please select one of them.

c-bata replied:
Thanks! I renamed it to `_to_intermediate_value_from_sqlalchemy_model` at bc1af1bc8.

toshihikoyanase commented on tests/storages_tests/rdb_tests/test_storage.py:
```diff
@@ -278,24 +280,6 @@ def test_upgrade_distributions(optuna_version: str) -> None:
         storage_url = f"sqlite:///{workdir}/sqlite.db"
 
         storage = RDBStorage(storage_url, skip_compatibility_check=True)
-        old_study = load_study(storage=storage, study_name="schema migration")
-        old_distribution_dict = old_study.trials[0].distributions
-
-        if version.parse(optuna_version) >= version.parse("3.0.0.a"):
-            assert isinstance(old_distribution_dict["x1"], FloatDistribution)
-            assert isinstance(old_distribution_dict["x2"], FloatDistribution)
-            assert isinstance(old_distribution_dict["x3"], FloatDistribution)
-            assert isinstance(old_distribution_dict["y1"], IntDistribution)
-            assert isinstance(old_distribution_dict["y2"], IntDistribution)
-            assert isinstance(old_distribution_dict["z"], CategoricalDistribution)
-        else:
-            assert isinstance(old_distribution_dict["x1"], UniformDistribution)
-            assert isinstance(old_distribution_dict["x2"], LogUniformDistribution)
-            assert isinstance(old_distribution_dict["x3"], DiscreteUniformDistribution)
-            assert isinstance(old_distribution_dict["y1"], IntUniformDistribution)
-            assert isinstance(old_distribution_dict["y2"], IntLogUniformDistribution)
-            assert isinstance(old_distribution_dict["z"], CategoricalDistribution)
```
I'm not sure if we can simply remove these assertions from the test. Let me revisit here the other day.
CC @himkt, I guess you wrote this test function. If you have any comments, please let us know.

toshihikoyanase replied:
I read the rest of the part of this test function, and it looks reasonable to skip the database contents check before upgrading them. As the comment above mentioned, the tables are not necessarily readable with the latest Optuna due to schema changes.

c-bata replied:
Yes. Thank you for your thorough review :bow:

toshihikoyanase replied:
Thank you for your update. Let me check the change with optuna-e2e, next.

toshihikoyanase reviewed (commented):
Thank you for your PR. I roughly checked the model definition and the unit tests. Let me share my early comments of them.

contramundum53 reviewed (approved):
LGTM except for Yanase-san's comments.

c-bata commented:
Thanks to the investigation by @toshihikoyanase, the migration script works as expected without some tricks. I rebased this branch on #3581 and removed `server_default` at  https://github.com/optuna/optuna/pull/3564/commits/66135bc443e878e4b01edff087823adf5abd946a :pray:

c-bata commented on optuna/storages/_rdb/alembic/versions/v3.0.0.c.py:
```diff
@@ -0,0 +1,123 @@
+
+
+def upgrade():
+    # MySQL and PostgreSQL supports DEFAULT clause like 'ALTER TABLE <tbl_name>
+    # ADD COLUMN <col_name> ... DEFAULT "FINITE_OR_NAN"', but seemingly Alembic
+    # does not support such a SQL statement. So first add a column with schema-level
+    # default value setting, then remove it by `batch_op.alter_column()`.
+    with op.batch_alter_table("trial_intermediate_values") as batch_op:
+        batch_op.add_column(
+            sa.Column(
+                "intermediate_value_type",
+                sa.Enum("FINITE_OR_NAN", "INF_POS", "INF_NEG", name="floattypeenum"),
+                nullable=False,
+                server_default="FINITE_OR_NAN",
+            ),
+        )
+    with op.batch_alter_table("trial_intermediate_values") as batch_op:
+        batch_op.alter_column("intermediate_value_type", server_default=None)
```
We actually need to execute these operations in separated batch (it means SQLite3 database recreated twice), otherwise `(sqlite3.IntegrityError) NOT NULL constraint failed: _alembic_tmp_trial_intermediate_values.intermediate_value_type` exception will be raised.

c-bata commented:
FYI: Now I'm investigating the following error that is found by running optuna-e2e tests on PostgreSQL.

```
$ docker compose run --rm optuna-300c bash src/upgrade.sh
mysql
[I 2022-05-24 02:56:38,223] Upgrading the storage schema to the latest version.
[I 2022-05-24 02:56:38,560] Completed to upgrade the storage.
postgresql
[I 2022-05-24 02:56:43,205] Upgrading the storage schema to the latest version.
[E 2022-05-24 02:56:43,465] (psycopg2.errors.UndefinedObject) type "floattypeenum" does not exist
LINE 1: ...mediate_values ADD COLUMN intermediate_value_type floattypee...
                                                             ^

[SQL: ALTER TABLE trial_intermediate_values ADD COLUMN intermediate_value_type floattypeenum DEFAULT 'FINITE_OR_NAN' NOT NULL]
(Background on this error at: https://sqlalche.me/e/14/f405)
sqlite
[I 2022-05-24 02:56:48,040] Upgrading the storage schema to the latest version.
[I 2022-05-24 02:56:48,595] Completed to upgrade the storage.
mssql
[I 2022-05-24 02:56:53,239] Upgrading the storage schema to the latest version.
[I 2022-05-24 02:56:53,513] Completed to upgrade the storage
```

c-bata commented:
I rebased on master branch since #3581 was merged.

c-bata commented on optuna/storages/_rdb/alembic/versions/v3.0.0.c.py:
```diff
@@ -0,0 +1,136 @@
+    intermediate_value_type = sa.Column(sa.Enum(FloatTypeEnum), nullable=False)
+
+
+def _get_postgresql_float_type_enum():
+    from sqlalchemy.dialects import postgresql
+
+    return postgresql.ENUM("FINITE_OR_NAN", "INF_POS", "INF_NEG", name="floattypeenum")
+
+
+def upgrade():
+    bind = op.get_bind()
+
+    if bind.dialect.name == "postgresql":
+        _get_postgresql_float_type_enum().create(bind)
```
`batch_op.add_column(sa.Column("...", sa.Enum(...)))` does not create Enum type on PostgreSQL, so that optuna-e2e script raised following errors.

```
$ docker compose run --rm optuna-300c bash src/upgrade.sh
mysql
...
postgresql
[I 2022-05-24 02:56:43,205] Upgrading the storage schema to the latest version.
[E 2022-05-24 02:56:43,465] (psycopg2.errors.UndefinedObject) type "floattypeenum" does not exist
LINE 1: ...mediate_values ADD COLUMN intermediate_value_type floattypee...
                                                             ^

[SQL: ALTER TABLE trial_intermediate_values ADD COLUMN intermediate_value_type floattypeenum DEFAULT 'FINITE_OR_NAN' NOT NULL]
(Background on this error at: https://sqlalche.me/e/14/f405)
...
```

<details>

<summary>list of Enum types</summary>

```
optuna=# alter table trial_intermediate_values add column intermediate_value_type floattypeenum default 'FINITE_OR_NAN' not null;
ERROR:  type "floattypeenum" does not exist
LINE 1: ...mediate_values add column intermediate_value_type floattypee...
                                                             ^
optuna=# select n.nspname as enum_schema,
       t.typname as enum_name,
       e.enumlabel as enum_value
from pg_type t
   join pg_enum e on t.oid = e.enumtypid
   join pg_catalog.pg_namespace n ON n.oid = t.typnamespace;
 enum_schema |   enum_name    | enum_value
-------------+----------------+------------
 public      | studydirection | NOT_SET
 public      | studydirection | MINIMIZE
 public      | studydirection | MAXIMIZE
 public      | trialstate     | RUNNING
 public      | trialstate     | COMPLETE
 public      | trialstate     | PRUNED
 public      | trialstate     | FAIL
 public      | trialstate     | WAITING

(8 rows)
```

</details>

According to the [SQLAlchemy's documentation](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#sqlalchemy.dialects.postgresql.ENUM), create/drop behavior of ENUM is necessarily intricate. In fact, when using Enum type in PostgreSQL, it is required to create Enum type in separated SQL as not like MySQL.

```
# PostgreSQL
CREATE TYPE floattypeenum AS ENUM ('FINITE_OR_NAN', 'INF_POS', 'INF_NEG')
ALTER TABLE trial_intermediate_values ADD COLUMN intermediate_value_type floattypeenum DEFAULT 'FINITE_OR_NAN' NOT NULL
```

```
# MySQL
ALTER TABLE trial_intermediate_values ADD COLUMN intermediate_value_type ENUM('FINITE_OR_NAN','INF_POS','INF_NEG') NOT NULL DEFAULT 'FINITE_OR_NAN'
```

For these reasons, our migration script requires the postgresql specific logics. As an alternative approach, we can use VARCHAR column with CHECK constraints by using [native_enum=False](https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum.params.native_enum) and [create_constraint=True](https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum.params.create_constraint) options.

c-bata replied:
@toshihikoyanase suggested me to use `sa.Enum(...).create()` method. Thanks! I'll try to use this.

* https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum.create
* https://github.com/sqlalchemy/alembic/issues/278#issuecomment-904343421

c-bata replied:
I confirmed that `sq.Enum(...).create()` works as expected. Full output of `bash src/upgrade.sh` of optuna-e2e with SQLAlchemy's INFO logging is here.

```
# bash src/upgrade.sh
mysql
[I 2022-05-24 08:29:01,607] Upgrading the storage schema to the latest version.
[I 2022-05-24 08:29:01,633] SELECT DATABASE()
[I 2022-05-24 08:29:01,634] [raw sql] {}
[I 2022-05-24 08:29:01,637] SELECT @@sql_mode
[I 2022-05-24 08:29:01,638] [raw sql] {}
[I 2022-05-24 08:29:01,640] SELECT @@lower_case_table_names
[I 2022-05-24 08:29:01,641] [raw sql] {}
[I 2022-05-24 08:29:01,647] SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = %(table_schema)s AND table_name = %(table_name)s
[I 2022-05-24 08:29:01,648] [generated in 0.00093s] {'table_schema': 'optuna', 'table_name': 'alembic_version'}
[I 2022-05-24 08:29:01,657] SELECT alembic_version.version_num
FROM alembic_version
[I 2022-05-24 08:29:01,658] [generated in 0.00111s] {}
[I 2022-05-24 08:29:01,752] BEGIN (implicit)
[I 2022-05-24 08:29:01,757] ALTER TABLE trial_intermediate_values ADD COLUMN intermediate_value_type ENUM('FINITE_OR_NAN','INF_POS','INF_NEG') NOT NULL DEFAULT 'FINITE_OR_NAN'
[I 2022-05-24 08:29:01,758] [no key 0.00076s] {}
[I 2022-05-24 08:29:01,801] ALTER TABLE trial_intermediate_values ALTER COLUMN intermediate_value_type DROP DEFAULT
[I 2022-05-24 08:29:01,801] [no key 0.00092s] {}
[I 2022-05-24 08:29:01,830] SELECT trial_intermediate_values.trial_intermediate_value_id AS trial_intermediate_values_trial_intermediate_value_id, trial_intermediate_values.intermediate_value AS trial_intermediate_values_intermediate_value, trial_intermediate_values.intermediate_value_type AS trial_intermediate_values_intermediate_value_type
FROM trial_intermediate_values
[I 2022-05-24 08:29:01,830] [generated in 0.00101s] {}
[I 2022-05-24 08:29:01,869] UPDATE trial_intermediate_values SET intermediate_value_type=%(intermediate_value_type)s WHERE trial_intermediate_values.trial_intermediate_value_id = %(trial_intermediate_values_trial_intermediate_value_id)s
[I 2022-05-24 08:29:01,870] [generated in 0.00169s] ({'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 1}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 2}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 4}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 5}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 7}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 8}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 10}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 11}  ... displaying 10 of 20 total bound parameter sets ...  {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 28}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 29})
[I 2022-05-24 08:29:01,903] UPDATE alembic_version SET version_num='v3.0.0.c' WHERE alembic_version.version_num = 'v3.0.0.b'
[I 2022-05-24 08:29:01,904] [generated in 0.00104s] {}
[I 2022-05-24 08:29:01,907] COMMIT
[I 2022-05-24 08:29:01,914] BEGIN (implicit)
[I 2022-05-24 08:29:01,916] SELECT version_info.version_info_id AS version_info_version_info_id, version_info.schema_version AS version_info_schema_version, version_info.library_version AS version_info_library_version
FROM version_info
[I 2022-05-24 08:29:01,916] [cached since 0.8519s ago] {}
[I 2022-05-24 08:29:01,923] COMMIT
[I 2022-05-24 08:29:01,926] Completed to upgrade the storage.
postgresql
[I 2022-05-24 08:29:05,792] Upgrading the storage schema to the latest version.
[I 2022-05-24 08:29:05,880] select pg_catalog.version()
[I 2022-05-24 08:29:05,881] [raw sql] {}
[I 2022-05-24 08:29:05,886] select current_schema()
[I 2022-05-24 08:29:05,886] [raw sql] {}
[I 2022-05-24 08:29:05,890] show standard_conforming_strings
[I 2022-05-24 08:29:05,891] [raw sql] {}
[I 2022-05-24 08:29:05,894] BEGIN (implicit)
[I 2022-05-24 08:29:05,896] select relname from pg_class c join pg_namespace n on n.oid=c.relnamespace where pg_catalog.pg_table_is_visible(c.oid) and relname=%(name)s
[I 2022-05-24 08:29:05,897] [generated in 0.00083s] {'name': 'alembic_version'}
[I 2022-05-24 08:29:05,905] SELECT alembic_version.version_num
FROM alembic_version
[I 2022-05-24 08:29:05,906] [generated in 0.00118s] {}
[I 2022-05-24 08:29:05,994]
            SELECT EXISTS (
                SELECT * FROM pg_catalog.pg_type t
                WHERE t.typname = %(typname)s
                AND pg_type_is_visible(t.oid)
                )

[I 2022-05-24 08:29:05,995] [generated in 0.00105s] {'typname': 'floattypeenum'}
[I 2022-05-24 08:29:06,003] CREATE TYPE floattypeenum AS ENUM ('FINITE_OR_NAN', 'INF_POS', 'INF_NEG')
[I 2022-05-24 08:29:06,003] [no key 0.00078s] {}
[I 2022-05-24 08:29:06,018] ALTER TABLE trial_intermediate_values ADD COLUMN intermediate_value_type floattypeenum DEFAULT 'FINITE_OR_NAN' NOT NULL
[I 2022-05-24 08:29:06,019] [no key 0.00086s] {}
[I 2022-05-24 08:29:06,031] ALTER TABLE trial_intermediate_values ALTER COLUMN intermediate_value_type DROP DEFAULT
[I 2022-05-24 08:29:06,032] [no key 0.00104s] {}
[I 2022-05-24 08:29:06,045] SELECT trial_intermediate_values.trial_intermediate_value_id AS trial_intermediate_values_trial_intermediate_value_id, trial_intermediate_values.intermediate_value AS trial_intermediate_values_intermediate_value, trial_intermediate_values.intermediate_value_type AS trial_intermediate_values_intermediate_value_type
FROM trial_intermediate_values
[I 2022-05-24 08:29:06,046] [generated in 0.00110s] {}
[I 2022-05-24 08:29:06,080] UPDATE trial_intermediate_values SET intermediate_value_type=%(intermediate_value_type)s WHERE trial_intermediate_values.trial_intermediate_value_id = %(trial_intermediate_values_trial_intermediate_value_id)s
[I 2022-05-24 08:29:06,081] [generated in 0.00179s] ({'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 1}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 2}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 4}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 5}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 7}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 8}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 10}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 11}  ... displaying 10 of 20 total bound parameter sets ...  {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 28}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 29})
[I 2022-05-24 08:29:06,106] UPDATE alembic_version SET version_num='v3.0.0.c' WHERE alembic_version.version_num = 'v3.0.0.b'
[I 2022-05-24 08:29:06,107] [generated in 0.00107s] {}
[I 2022-05-24 08:29:06,113] COMMIT
[I 2022-05-24 08:29:06,119] BEGIN (implicit)
[I 2022-05-24 08:29:06,120] SELECT version_info.version_info_id AS version_info_version_info_id, version_info.schema_version AS version_info_schema_version, version_info.library_version AS version_info_library_version
FROM version_info
[I 2022-05-24 08:29:06,121] [cached since 0.7972s ago] {}
[I 2022-05-24 08:29:06,126] COMMIT
[I 2022-05-24 08:29:06,128] Completed to upgrade the storage.
sqlite
[I 2022-05-24 08:29:09,960] Upgrading the storage schema to the latest version.
[I 2022-05-24 08:29:09,984] PRAGMA main.table_info("alembic_version")
[I 2022-05-24 08:29:09,985] [raw sql] ()
[I 2022-05-24 08:29:09,995] SELECT alembic_version.version_num
FROM alembic_version
[I 2022-05-24 08:29:09,996] [generated in 0.00098s] ()
[I 2022-05-24 08:29:10,091] BEGIN (implicit)
[I 2022-05-24 08:29:10,096] ALTER TABLE trial_intermediate_values ADD COLUMN intermediate_value_type VARCHAR(13) DEFAULT 'FINITE_OR_NAN' NOT NULL
[I 2022-05-24 08:29:10,097] [no key 0.00079s] ()
[I 2022-05-24 08:29:10,138] PRAGMA main.table_info("trial_intermediate_values")
[I 2022-05-24 08:29:10,139] [raw sql] ()
[I 2022-05-24 08:29:10,151] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,152] [raw sql] ('trial_intermediate_values',)
[I 2022-05-24 08:29:10,160] PRAGMA main.foreign_key_list("trial_intermediate_values")
[I 2022-05-24 08:29:10,161] [raw sql] ()
[I 2022-05-24 08:29:10,166] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,166] [raw sql] ('trial_intermediate_values',)
[I 2022-05-24 08:29:10,182] PRAGMA main.table_info("trials")
[I 2022-05-24 08:29:10,183] [raw sql] ()
[I 2022-05-24 08:29:10,191] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,192] [raw sql] ('trials',)
[I 2022-05-24 08:29:10,197] PRAGMA main.foreign_key_list("trials")
[I 2022-05-24 08:29:10,197] [raw sql] ()
[I 2022-05-24 08:29:10,202] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,204] [raw sql] ('trials',)
[I 2022-05-24 08:29:10,210] PRAGMA main.table_info("studies")
[I 2022-05-24 08:29:10,211] [raw sql] ()
[I 2022-05-24 08:29:10,217] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,218] [raw sql] ('studies',)
[I 2022-05-24 08:29:10,222] PRAGMA main.foreign_key_list("studies")
[I 2022-05-24 08:29:10,223] [raw sql] ()
[I 2022-05-24 08:29:10,225] PRAGMA temp.foreign_key_list("studies")
[I 2022-05-24 08:29:10,225] [raw sql] ()
[I 2022-05-24 08:29:10,227] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,228] [raw sql] ('studies',)
[I 2022-05-24 08:29:10,232] PRAGMA main.index_list("studies")
[I 2022-05-24 08:29:10,233] [raw sql] ()
[I 2022-05-24 08:29:10,238] PRAGMA main.index_info("ix_studies_study_name")
[I 2022-05-24 08:29:10,239] [raw sql] ()
[I 2022-05-24 08:29:10,244] PRAGMA main.index_list("studies")
[I 2022-05-24 08:29:10,245] [raw sql] ()
[I 2022-05-24 08:29:10,249] PRAGMA main.index_info("ix_studies_study_name")
[I 2022-05-24 08:29:10,250] [raw sql] ()
[I 2022-05-24 08:29:10,254] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,254] [raw sql] ('studies',)
[I 2022-05-24 08:29:10,270] PRAGMA main.index_list("trials")
[I 2022-05-24 08:29:10,271] [raw sql] ()
[I 2022-05-24 08:29:10,275] PRAGMA temp.index_list("trials")
[I 2022-05-24 08:29:10,276] [raw sql] ()
[I 2022-05-24 08:29:10,278] PRAGMA main.index_list("trials")
[I 2022-05-24 08:29:10,278] [raw sql] ()
[I 2022-05-24 08:29:10,283] PRAGMA temp.index_list("trials")
[I 2022-05-24 08:29:10,283] [raw sql] ()
[I 2022-05-24 08:29:10,285] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,286] [raw sql] ('trials',)
[I 2022-05-24 08:29:10,292] PRAGMA main.index_list("trial_intermediate_values")
[I 2022-05-24 08:29:10,293] [raw sql] ()
[I 2022-05-24 08:29:10,298] PRAGMA main.index_list("trial_intermediate_values")
[I 2022-05-24 08:29:10,299] [raw sql] ()
[I 2022-05-24 08:29:10,303] PRAGMA main.index_info("sqlite_autoindex_trial_intermediate_values_1")
[I 2022-05-24 08:29:10,304] [raw sql] ()
[I 2022-05-24 08:29:10,308] SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
[I 2022-05-24 08:29:10,309] [raw sql] ('trial_intermediate_values',)
[I 2022-05-24 08:29:10,323]
CREATE TABLE _alembic_tmp_trial_intermediate_values (
	trial_intermediate_value_id INTEGER NOT NULL,
	trial_id INTEGER NOT NULL,
	step INTEGER NOT NULL,
	intermediate_value FLOAT,
	intermediate_value_type VARCHAR(13) NOT NULL,
	PRIMARY KEY (trial_intermediate_value_id),
	UNIQUE (trial_id, step),
	FOREIGN KEY(trial_id) REFERENCES trials (trial_id)
)


[I 2022-05-24 08:29:10,324] [no key 0.00097s] ()
[I 2022-05-24 08:29:10,359] INSERT INTO _alembic_tmp_trial_intermediate_values (trial_intermediate_value_id, trial_id, step, intermediate_value, intermediate_value_type) SELECT trial_intermediate_values.trial_intermediate_value_id, trial_intermediate_values.trial_id, trial_intermediate_values.step, trial_intermediate_values.intermediate_value, trial_intermediate_values.intermediate_value_type
FROM trial_intermediate_values
[I 2022-05-24 08:29:10,360] [generated in 0.00112s] ()
[I 2022-05-24 08:29:10,388]
DROP TABLE trial_intermediate_values
[I 2022-05-24 08:29:10,389] [no key 0.00109s] ()
[I 2022-05-24 08:29:10,408] ALTER TABLE _alembic_tmp_trial_intermediate_values RENAME TO trial_intermediate_values
[I 2022-05-24 08:29:10,409] [no key 0.00098s] ()
[I 2022-05-24 08:29:10,424] SELECT trial_intermediate_values.trial_intermediate_value_id AS trial_intermediate_values_trial_intermediate_value_id, trial_intermediate_values.intermediate_value AS trial_intermediate_values_intermediate_value, trial_intermediate_values.intermediate_value_type AS trial_intermediate_values_intermediate_value_type
FROM trial_intermediate_values
[I 2022-05-24 08:29:10,425] [generated in 0.00096s] ()
[I 2022-05-24 08:29:10,456] UPDATE trial_intermediate_values SET intermediate_value_type=? WHERE trial_intermediate_values.trial_intermediate_value_id = ?
[I 2022-05-24 08:29:10,456] [generated in 0.00163s] (('INF_POS', 1), ('INF_NEG', 2), ('INF_POS', 4), ('INF_NEG', 5), ('INF_POS', 7), ('INF_NEG', 8), ('INF_POS', 10), ('INF_NEG', 11)  ... displaying 10 of 20 total bound parameter sets ...  ('INF_POS', 28), ('INF_NEG', 29))
[I 2022-05-24 08:29:10,464] UPDATE alembic_version SET version_num='v3.0.0.c' WHERE alembic_version.version_num = 'v3.0.0.b'
[I 2022-05-24 08:29:10,465] [generated in 0.00095s] ()
[I 2022-05-24 08:29:10,479] COMMIT
[I 2022-05-24 08:29:10,500] BEGIN (implicit)
[I 2022-05-24 08:29:10,502] SELECT version_info.version_info_id AS version_info_version_info_id, version_info.schema_version AS version_info_schema_version, version_info.library_version AS version_info_library_version
FROM version_info
[I 2022-05-24 08:29:10,503] [cached since 1.084s ago] ()
[I 2022-05-24 08:29:10,518] COMMIT
[I 2022-05-24 08:29:10,520] Completed to upgrade the storage.
mssql
[I 2022-05-24 08:29:14,362] Upgrading the storage schema to the latest version.
[I 2022-05-24 08:29:14,403] select @@version
[I 2022-05-24 08:29:14,404] [raw sql] {}
[I 2022-05-24 08:29:14,407] SELECT schema_name()
[I 2022-05-24 08:29:14,408] [generated in 0.00091s] {}
[I 2022-05-24 08:29:14,414] SELECT CAST('test max support' AS NVARCHAR(max))
[I 2022-05-24 08:29:14,415] [generated in 0.00109s] {}
[I 2022-05-24 08:29:14,421] BEGIN (implicit)
[I 2022-05-24 08:29:14,429] SELECT [INFORMATION_SCHEMA].[TABLES].[TABLE_NAME]
FROM [INFORMATION_SCHEMA].[TABLES]
WHERE [INFORMATION_SCHEMA].[TABLES].[TABLE_TYPE] = CAST(%(table_type_1)s AS NVARCHAR(max)) AND [INFORMATION_SCHEMA].[TABLES].[TABLE_NAME] = CAST(%(table_name_1)s AS NVARCHAR(max)) AND [INFORMATION_SCHEMA].[TABLES].[TABLE_SCHEMA] = CAST(%(table_schema_1)s AS NVARCHAR(max))
[I 2022-05-24 08:29:14,430] [generated in 0.00132s] {'table_type_1': 'BASE TABLE', 'table_name_1': 'alembic_version', 'table_schema_1': 'dbo'}
[I 2022-05-24 08:29:14,436] SELECT alembic_version.version_num
FROM alembic_version
[I 2022-05-24 08:29:14,436] [generated in 0.00113s] {}
[I 2022-05-24 08:29:14,531] ALTER TABLE trial_intermediate_values ADD intermediate_value_type VARCHAR(13) NOT NULL DEFAULT 'FINITE_OR_NAN'
[I 2022-05-24 08:29:14,532] [no key 0.00087s] {}
[I 2022-05-24 08:29:14,538] declare @const_name varchar(256)
select @const_name = [name] from sys.default_constraints
where parent_object_id = object_id('trial_intermediate_values')
and col_name(parent_object_id, parent_column_id) = 'intermediate_value_type'
exec('alter table trial_intermediate_values drop constraint ' + @const_name)
[I 2022-05-24 08:29:14,538] [no key 0.00082s] {}
[I 2022-05-24 08:29:14,563] SELECT trial_intermediate_values.trial_intermediate_value_id AS trial_intermediate_values_trial_intermediate_value_id, trial_intermediate_values.intermediate_value AS trial_intermediate_values_intermediate_value, trial_intermediate_values.intermediate_value_type AS trial_intermediate_values_intermediate_value_type
FROM trial_intermediate_values
[I 2022-05-24 08:29:14,564] [generated in 0.00102s] {}
[I 2022-05-24 08:29:14,596] UPDATE trial_intermediate_values SET intermediate_value_type=%(intermediate_value_type)s WHERE trial_intermediate_values.trial_intermediate_value_id = %(trial_intermediate_values_trial_intermediate_value_id)s
[I 2022-05-24 08:29:14,597] [generated in 0.00168s] ({'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 1}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 2}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 4}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 5}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 7}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 8}, {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 10}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 11}  ... displaying 10 of 20 total bound parameter sets ...  {'intermediate_value_type': 'INF_POS', 'trial_intermediate_values_trial_intermediate_value_id': 28}, {'intermediate_value_type': 'INF_NEG', 'trial_intermediate_values_trial_intermediate_value_id': 29})
[I 2022-05-24 08:29:14,616] UPDATE alembic_version SET version_num='v3.0.0.c' WHERE alembic_version.version_num = 'v3.0.0.b'
[I 2022-05-24 08:29:14,617] [generated in 0.00098s] {}
[I 2022-05-24 08:29:14,621] COMMIT
[I 2022-05-24 08:29:14,630] BEGIN (implicit)
[I 2022-05-24 08:29:14,631] SELECT version_info.version_info_id AS version_info_version_info_id, version_info.schema_version AS version_info_schema_version, version_info.library_version AS version_info_library_version
FROM version_info
[I 2022-05-24 08:29:14,632] [cached since 0.8021s ago] {}
[I 2022-05-24 08:29:14,638] COMMIT
[I 2022-05-24 08:29:14,641] Completed to upgrade the storage.
```

contramundum53 commented:
Should I review this again?

c-bata commented:
@contramundum53 Thank you for the confirmation. Yes, this PR is ready for review. Please proceed to review again :bow:

contramundum53 commented on optuna/storages/_rdb/alembic/versions/v3.0.0.c.py:
```diff
@@ -0,0 +1,128 @@
+                continue
+
+            _intermediate_value = r.intermediate_value
+            if r.intermediate_value_type == IntermediateValueModel.FloatTypeEnum.INF_POS:
+                _intermediate_value = RDB_MAX_FLOAT
+            else:
+                _intermediate_value = RDB_MIN_FLOAT
```
FYI, for databases that accept `inf` values, migrating from pre-v3 to v3 and then downgrading changes the values. If we don't want this behavior, we could throw away downgrade support altogether.

contramundum53 commented:
I haven't tested it with optuna-e2e but the changes look good to me.

contramundum53 reviewed (approved):
Confirmed. LGTM.

toshihikoyanase commented on optuna/storages/_rdb/storage.py:
```diff
@@ -793,6 +793,7 @@ def _ensure_not_nan(value: float) -> Optional[float]:
 
     @staticmethod
     def _ensure_numerical_limit(value: float) -> float:
+        # TODO(c-bata): Remove this method after fixing inf/-inf handling of trial.values
```
A nit.
```suggestion
        # TODO(c-bata): Remove this method after fixing inf/-inf handling of trial.values.
```

toshihikoyanase commented on optuna/storages/_rdb/storage.py:
```diff
@@ -802,6 +803,7 @@ def _ensure_numerical_limit(value: float) -> float:
 
     @staticmethod
     def _lift_numerical_limit(value: Optional[float]) -> float:
+        # TODO(c-bata): Remove this method after fixing inf/-inf handling of trial.values
```
ditto.
```suggestion
        # TODO(c-bata): Remove this method after fixing inf/-inf handling of trial.values.
```

toshihikoyanase commented on optuna/storages/_rdb/alembic/versions/v3.0.0.c.py:
```diff
@@ -0,0 +1,128 @@
+        mapping = []
+        for r in records:
+            float_type: IntermediateValueModel.FloatTypeEnum
+            if np.isclose(r.intermediate_value, RDB_MAX_FLOAT) or np.isposinf(
+                r.intermediate_value
+            ):
```
[Notes] I thought the value clipping would be applied if users used MySQL backend, but actually not. The clipping is always applied no matter which backend users use.
https://github.com/optuna/optuna/blob/cbbcaa1d676e0455b97e84934ece44b14a035371/optuna/storages/_rdb/storage.py#L861

contramundum53 replied:
@toshihikoyanase FIY, this code is for migrating from pre-v3 to v3 with non-MySQL backend, where inf values are stored as-is.

toshihikoyanase reviewed (commented):
Thank you for your update.
I confirmed the change with optuna-e2e and worked as expected.

The change basically looks good to me. I added two small suggestions. Could you check them, please?

<details>
<summary>Reproducible code</summary>

init.py

```python
import optuna
import pymssql

from objective import objective


def create_mssql_database() -> None:
    conn = pymssql.connect("mssql", "sa", "optuna-test-5ZYB")
    conn.autocommit(True)
    cursor = conn.cursor()
    cursor.execute("CREATE DATABASE optuna")


create_mssql_database()


print("mysql")
study1 = optuna.create_study(study_name="test", storage="mysql+pymysql://root:root@mysql:3306/optuna")
try:
    study1.optimize(objective, n_trials=10)
except:
    pass

print("postgresql")
study2 = optuna.create_study(study_name="test", storage="postgresql+psycopg2://root:root@postgresql/optuna")
try:
    study2.optimize(objective, n_trials=10)
except:
    pass
print("sqlite")
study3 = optuna.create_study(study_name="test", storage="sqlite:///data/sample.db")
try:
    study3.optimize(objective, n_trials=10)
except:
    pass
print("mssql")
study4 = optuna.create_study(study_name="test", storage="mssql+pymssql://sa:optuna-test-5ZYB@mssql/optuna?charset=utf8")
try:
    study4.optimize(objective, n_trials=10)
except:
    pass
print("redis")
study5 = optuna.create_study(study_name="test", storage="redis://redis:6379")
try:
    study5.optimize(objective, n_trials=10)
except:
    pass

print(study1)
print(study2)
print(study3)
print(study4)
print(study5)

print("done")
```

load.py

```python
import optuna

from objective import objective


print("mysql")
study1 = optuna.load_study(study_name="test", storage="mysql+pymysql://root:root@mysql:3306/optuna")
try:
    study1.optimize(objective, n_trials=1)
except:
    pass
print("postgresql")
study2 = optuna.load_study(study_name="test", storage="postgresql+psycopg2://root:root@postgresql/optuna")
try:
    study2.optimize(objective, n_trials=1)
except:
    pass
print("sqlite")
study3 = optuna.load_study(study_name="test", storage="sqlite:///data/sample.db")
try:
    study3.optimize(objective, n_trials=1)
except:
    pass
print("mssql")
study4 = optuna.load_study(study_name="test", storage="mssql+pymssql://sa:optuna-test-5ZYB@mssql/optuna?charset=utf8")
try:
    study4.optimize(objective, n_trials=1)
except:
    pass
print("redis")
study5 = optuna.load_study(study_name="test", storage="redis://redis:6379")
try:
    study5.optimize(objective, n_trials=1)
except:
    pass

print(study1.trials[-1].intermediate_values)
print(study2.trials[-1].intermediate_values)
print(study3.trials[-1].intermediate_values)
print(study4.trials[-1].intermediate_values)
print(study5.trials[-1].intermediate_values)

print("done")
```

objective.py

```python
import optuna


def objective(trial: optuna.Trial) -> float:
    x1 = trial.suggest_float("x1", 0, 10)
    x2 = trial.suggest_float("x2", 1, 10, log=True)
    x3 = trial.suggest_float("x3", 0, 10, step=2.0)
    y1 = trial.suggest_int("y1", 0, 10)
    y2 = trial.suggest_int("y2", 1, 10, log=True)
    y3 = trial.suggest_int("y3", 0, 10, step=2)
    c1 = trial.suggest_categorical("c1", [1, 10, 100])
    trial.report(value=1, step=1)
    trial.report(value=float("inf"), step=2)
    trial.report(value=-float("inf"), step=3)
    trial.report(value=-float("nan"), step=4)
    trial.report(value=None, step=5)
    return x1**y1 + x2**y2 + x3**y3 + c1
```

optuna-300/Dockerfile

```
FROM python:3.9.10-buster

RUN pip install -U pip
RUN pip install git+https://github.com/c-bata/optuna@fix-intermediate-value-inf
RUN pip install PyMySQL psycopg2-binary cryptography redis pymssql

WORKDIR /work
```

</details>

Optuna v2.10.0 does not support inf/nan as follows, but this feature branch correctly handled them.

Optuna v2.10.0

| DB | 1 | inf | -inf | nan |
| --- | --- | --- | --- | --- |
| mysql | OK | NG | ? | ? |
| postgres | OK | OK | OK | OK |
| sqlite | OK | OK | OK | NG | (not null)
| mssql | OK | NG | ? | ? |
| redis | OK | OK | OK | OK |

- MySQL and MsSQL do not support `inf`
- `sqlite` does not distinguish `nan` and `None`, so `nan` was rejected due to `nullable=False`

c-bata commented:
Thanks! I applied your suggestions :bow:

toshihikoyanase reviewed (approved):
Thank you for your update. LGTM!

toshihikoyanase commented:
Let me slightly update the title for the release note.
